crmsh/utils.py

```diff
@@ -2292,4 +2149,255 @@ def interface_choice():
+        """
+        cmd = "ip -{} -o addr show".format(self.ip_version)
+        rc, out, err = get_stdout_stderr(cmd)
+        if rc != 0:
+            raise ValueError(err)
+
+        # format on each line will like:
+        # 2: enp1s0    inet 192.168.122.241/24 brd 192.168.122.255 scope global enp1s0\       valid_lft forever preferred_lft forever
+        for line in out.splitlines():
+            _, nic, _, ip_with_mask, *_ = line.split()
+            #TODO change this condition when corosync support link-local address
+            interface_inst = Interface(ip_with_mask)
+            if interface_inst.is_loopback or interface_inst.is_link_local:
+                continue
+            # one nic might configured multi IP addresses
+            if nic not in self._nic_info_dict:
+                self._nic_info_dict[nic] = []
+            self._nic_info_dict[nic].append(interface_inst)
+
+        if not self._nic_info_dict:
+            raise ValueError("No address configured")
+        if self.second_heartbeat and len(self._nic_info_dict) == 1:
+            raise ValueError("Cannot configure second heartbeat, since only one address is available")
+
+    @property
+    def nic_list(self):
+        """
+        Get interfaces name list
+        """
+        return list(self._nic_info_dict.keys())
+
+    @property
+    def interface_list(self):
+        """
+        Get instance list of class Interface
+        """
+        _interface_list = []
+        for interface in self._nic_info_dict.values():
+            _interface_list.extend(interface)
+        return _interface_list
+
+    @property
+    def ip_list(self):
+        """
+        Get IP address list
+        """
+        return [interface.ip for interface in self.interface_list]
+
+    @classmethod
+    def get_local_ip_list(cls, is_ipv6):
+        """
+        Get IP address list
+        """
+        cls_inst = cls(is_ipv6)
+        cls_inst.get_interfaces_info()
+        return cls_inst.ip_list
+
+    @property
+    def network_list(self):
+        """
+        Get network list
+        """
+        return list(set([interface.network for interface in self.interface_list]))
+
+    def nic_first_ip(self, nic):
+        """
+        Get the first IP of specific nic
+        """
+        return self._nic_info_dict[nic][0].ip
+
+    def get_default_nic_list_from_route(self):
+        """
+        Get default nic list from route
+        """
+        if self._default_nic_list:
+            return self._default_nic_list
+
+        #TODO what if user only has ipv6 route?
+        cmd = "ip -o route show"
+        rc, out, err = get_stdout_stderr(cmd)
+        if rc != 0:
+            raise ValueError(err)
+        res = re.search(r'^default via .* dev (.*?) ', out)
+        if res:
+            self._default_nic_list = [res.group(1)]
+        else:
+            common_warn("No default route configured. Using the first found nic")
+            self._default_nic_list = [self.nic_list[0]]
```

Comment:
This will only work if `get_interfaces_info` is executed before, right?
So we might have 2 different returns depending on how we use this method, not so nice.
Should be run `get_interfaces_info` here just in case?

Reply:
Thanks, changed